File: 1074-number-of-submatrices-that-sum-to-target/1074-number-of-submatrices-that-sum-to-target.py
class Solution:
    def numSubmatrixSumTarget(self, mat: List[List[int]], target: int) -> int:
         
        R = len(mat) ; C = len(mat[0]) 
        for r in range(R):
            for c in range(C-1):
                 mat[r][c+1] += mat[r][c] 

        ans = 0 
        for y1 in range(C):
            for y2 in range(y1,C): 
                prSum = {0:1} 
                sv = 0 
                for x in range(R): 
                    sv += mat[x][y2] - (mat[x][y1-1] if y1>=1 else 0 ) 
                    ans += prSum.get(sv-target , 0) 
                    prSum[sv] = prSum.get(sv,0) +1 

        return ans 

File: 1074-number-of-submatrices-that-sum-to-target/1074-number-of-submatrices-that-sum-to-target.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& mat, int target) {
        
        
        int ans = 0 , R = mat.size() , C = mat[0].size() ; 
        for(int r= 0; r < R ; r++)
            for(int c = 1 ; c < C ; c++)
                mat[r][c] += mat[r][c-1] ; 

        unordered_map<int,int> cntr ; 
        for(int y1 =0; y1 < C ; y1++){
            for(int y2= y1 ; y2 < C ; y2++){
                cntr = {{0,1}}; 
                int sv = 0 ; 
                for(int x =0 ; x < R ; x++){
                    sv += mat[x][y2] - ( y1 >0 ? mat[x][y1-1] : 0) ; 
                    ans += cntr.find(sv-target) != cntr.end()? cntr[sv-target] : 0;
                    cntr[sv] ++ ; 
                }}}
        return ans ; 
    }
};